File: RegExpMatching.java
//Time Complexity O(M*N)
//Space Complexity O(M*N)
//Leetcode tested


public class RegExpMatching {
    public boolean isMatch(String s, String p) {
        int sl = s.length();
        int pl = p.length();
        boolean[][]dp = new boolean[sl+1][pl+1];
        dp[0][0] = true;
        for (int j = 1; j < dp[0].length; j++) {
            if(p.charAt(j-1) == '*'){
                dp[0][j] = dp[0][j-2];
            }
        }
        for (int i = 1; i < dp.length; i++) {
            for (int j = 1; j < dp[0].length; j++) {
                if(p.charAt(j-1) != '*'){
                    if(p.charAt(j-1) == s.charAt(i-1) || p.charAt(j-1) == '.'){
                        dp[i][j] = dp[i-1][j-1];
                    }
                }else{
                    dp[i][j] = dp[i][j-2];
                    if(p.charAt(j-2) == s.charAt(i-2) || p.charAt(j-2) == '.'){
                        dp[i][j] = dp[i][j] || dp[i-1][j];
                    }
                }
            }
        }
        return dp[dp.length-1][dp[0].length-1];
    }
}

File: regexpmatching.py
# Time Complexity : O(M*N)
# Space Complexity : O(M*N)
# Did this code successfully run on Leetcode : Yes
# Any problem you faced while coding this : No


# Your code here along with comments explaining your approach
class Solution:
    def isMatch(self, s: str, p: str) -> bool:
        
        m = len(s)
        n = len(p)
        
        dp = [[False]*(n+1) for i in range(m+1)]
        
        dp[0][0] = True
        
        for i in range(0,m+1):
            for j in range(1,n+1):
                
                if i == 0 and p[j-1] == '*':
                    dp[i][j] = dp[i][j-2]
                    if len(s) == 0 and j == n:
                        return dp[i][j]
                    continue
                if len(s) == 0 and i==0:
                    continue
                
                    
                if s[i-1] == p[j-1] or p[j-1] == '.':
                    dp[i][j] = dp[i-1][j-1]
                
                elif p[j-1] == '*':
                    dp[i][j] = dp[i][j-2]
                    
                    if p[j-2] == s[i-1] or p[j-2] == '.':
                        dp[i][j] = dp[i][j-2] or dp[i-1][j]
        print(dp)  
        return dp[m][n]
                    
        